indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentInsertAction.java

```diff
@@ -59,26 +72,41 @@ public SegmentInsertAction(
     return segments;
   }
 
-  public TypeReference<Set<DataSegment>> getReturnTypeReference()
+  public TypeReference<SegmentPublishResult> getReturnTypeReference()
   {
-    return new TypeReference<Set<DataSegment>>()
+    return new TypeReference<SegmentPublishResult>()
     {
     };
   }
 
+  /**
+   * Behaves similarly to
+   * {@link io.druid.indexing.overlord.IndexerMetadataStorageCoordinator#announceHistoricalSegments(Set, DataSourceMetadata, DataSourceMetadata)},
+   * including the possibility of returning null in case of metadata transaction failure.
+   */
   @Override
-  public Set<DataSegment> perform(Task task, TaskActionToolbox toolbox) throws IOException
+  public SegmentPublishResult perform(Task task, TaskActionToolbox toolbox) throws IOException
   {
     toolbox.verifyTaskLocks(task, segments);
 
-    final Set<DataSegment> retVal = toolbox.getIndexerMetadataStorageCoordinator().announceHistoricalSegments(segments);
+    final SegmentPublishResult retVal = toolbox.getIndexerMetadataStorageCoordinator().announceHistoricalSegments(
+        segments,
+        startMetadata,
+        endMetadata
+    );
 
     // Emit metrics
     final ServiceMetricEvent.Builder metricBuilder = new ServiceMetricEvent.Builder()
         .setDimension(DruidMetrics.DATASOURCE, task.getDataSource())
         .setDimension(DruidMetrics.TASK_TYPE, task.getType());
 
-    for (DataSegment segment : segments) {
+    if (retVal.isSuccess()) {
+      toolbox.getEmitter().emit(metricBuilder.build("segment/txn/success", 1));
+    } else {
+      toolbox.getEmitter().emit(metricBuilder.build("segment/txn/failure", 1));
+    }
```

Comment:
maybe you want to confirm that txns are actually happening? I dunno? It seemed like if one existed then the other should too